Goals cleanup

glesica opened the pull request:
This PR cleans up the `Goals.jl` code Alden wrote by making it more Julia-ish and in some cases, perhaps, a bit clearer and faster. It is a work-in-progress still, but this way we can look at it in a sane manner.

glesica committed:
Initial implementations.

glesica committed:
Add additional convert methods. Clean up constructors.

glesica committed:
Add the remaining convert implementations.

glesica committed:
Add print method for packed goals.

glesica committed:
Cleanup comments.

glesica committed:
Refactor compose function.

glesica committed:
Fix constructor type param issues.

glesica committed:
Fix inexact error.

glesica committed:
Fix inexact error.

glesica committed:
Really fix inexact error.

glesica commented:
I went ahead and fixed the tests so that we can get back to passing on both 0.3 and 0.4. Obviously we need to add some tests, but at least we'll start from a "clean" position.

glesica committed:
Add goal tests and fix bugs to make them pass.

glesica committed:
Add tests for invalid comparisons.

glesica commented on src/Func.jl:
```diff
@@ -13,7 +13,7 @@ const OR = Func(|, 2)
 const NOT = Func(~, 1)
 const ZERO = Func(() -> convert(BitString, 0), 0, "0")
-const ONE = Func(() -> convert(BitString, ~0), 0, "1")
+const ONE = Func(() -> ~convert(BitString, 0), 0, "1")
```
This was done because in Julia 0.4 the `~0` syntax resulted in a loss of precision error.

ahalwright replied:
I can work on adding tests of goal composition and fitness to
test/Goals.jl.  It may be tomorrow before I get anything done.  Do you
want to delay merging goals.cleanup until I get this done?

Alden

On Wed, Mar 11, 2015, at 12:52, George Lesica wrote:

> > @@ -13,7 +13,7 @@ const OR = Func(|, 2)
> >  const XOR = Func($, 2)
> >  const NOT = Func(~, 1)
> >  const ZERO = Func(() -> convert(BitString, 0), 0, "0")
> > -const ONE = Func(() -> convert(BitString, ~0), 0, "1")
> > +const ONE = Func(() -> ~convert(BitString, 0), 0, "1")
> 
> This was done because in Julia 0.4 the `~0` syntax resulted in a loss of
> precision error.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/glesica/CGP.jl/pull/4/files#r26243394

glesica replied:
Yeah, go ahead and add the tests to that branch and push it. I'll wait
to merge until we can have everything working nicely.

On Wed, Mar 11, 2015, at 05:46 PM, Alden Wright wrote:

> > @@ -13,7 +13,7 @@ const OR = Func(|, 2)
> >  const XOR = Func($, 2)
> >  const NOT = Func(~, 1)
> >  const ZERO = Func(() -> convert(BitString, 0), 0, "0")
> > -const ONE = Func(() -> convert(BitString, ~0), 0, "1")
> > +const ONE = Func(() -> ~convert(BitString, 0), 0, "1")
> 
> I can work on adding tests of goal composition and fitness to
> test/Goals.jl.  It may be tomorrow before I get anything done.  Do you
> want to delay merging goals.cleanup until I get this done?
> 
> Alden
> 
> On Wed, Mar 11, 2015, at 12:52, George Lesica wrote:
> 
> > > @@ -13,7 +13,7 @@ const OR = Func(|, 2)
> > >  const XOR = Func($, 2)
> > >  const NOT = Func(~, 1)
> > >  const ZERO = Func(() -> convert(BitString, 0), 0, "0")
> > > -const ONE = Func(() -> convert(BitString, ~0), 0, "1")
> > > +const ONE = Func(() -> ~convert(BitString, 0), 0, "1")
> > 
> > This was done because in Julia 0.4 the `~0` syntax resulted in a loss of
> > precision error.
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/glesica/CGP.jl/pull/4/files#r26243394
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/glesica/CGP.jl/pull/4/files#r26253805

glesica committed:
Remove useless todo.

glesica commented:
I'm thinking that we might want to put the fitness function and tests in a new file called `Fitness.jl` instead of in `Goal.jl` or `Chromosome.jl` since we could end up with several implementations. Having everything in one, predictable place would make things simpler for users and easier to maintain.

ahalwright committed:
Line 84:  multiplied 'sizeof(BitString)' by 8 to convert size to bits
line 162: function compose:  added '=' (assignment operator)
line 237:  changed return type to BasicPackedGoal to agree with current goal types.
Note:  I need to check that this is right.

ahalwright committed:
Added 3-input 3-output tests of conversion functions.
Added tests of execute.

glesica commented:
You'll need to include the *.plu file in the repo or the tests won't pass.

glesica committed:
Add fitness function.

glesica committed:
Use type parameter for goal.

glesica committed:
Add very basic fitness tests.

glesica commented:
I think we're in a good place with the goals stuff, so once we have a PLU file in there and the tests pass I'm going to merge this pull request.

glesica committed:
Add a todo to consolidate output_mask functions.

glesica committed:
Remove noisy print.

glesica committed:
Remove prints in tests and reformat a bit.

glesica committed:
Add convenience script for running tests.
